File: internal/utils/permissions.go
package utils

import (
	"discord-mod-bot/internal/config"
	"errors"
	"sync"
	"time"

	"github.com/bwmarrin/discordgo"
)

const (
	RoleAdmin = "admin"
	RoleMod   = "mod"
	RoleStaff = "staff"
)

var (
	modBanCounts  = make(map[string]map[string]int) // userID -> date -> count
	modKickCounts = make(map[string]map[string]int) // userID -> date -> count
	rateLimitMux  sync.RWMutex                      // Mutex for thread-safe access
	maxModActions = 10
)

// HasPermission checks if a user has the required permission level
// Optimized: Uses map lookup instead of multiple loops
func HasPermission(s *discordgo.Session, guildID, userID, requiredRole string) (bool, error) {
	// Try to get member from state cache first (faster)
	member, err := s.State.Member(guildID, userID)
	if err != nil {
		// Fallback to API call if not in cache
		member, err = s.GuildMember(guildID, userID)
		if err != nil {
			return false, err
		}
	}

	if member == nil {
		return false, nil
	}

	// Create map for O(1) lookup instead of O(n) loops
	roleMap := make(map[string]bool, len(member.Roles))
	for _, roleID := range member.Roles {
		roleMap[roleID] = true
	}

	// Check roles in priority order (admin > staff > mod)
	if config.Cfg.AdminRoleID != "" && roleMap[config.Cfg.AdminRoleID] {
		return true, nil
	}

	if config.Cfg.StaffRoleID != "" && roleMap[config.Cfg.StaffRoleID] {
		if requiredRole == RoleAdmin || requiredRole == RoleStaff {
			return true, nil
		}
	}

	if config.Cfg.ModRoleID != "" && roleMap[config.Cfg.ModRoleID] {
		if requiredRole == RoleMod {
			return true, nil
		}
	}

	return false, nil
}

// CanPerformModAction checks if a mod can perform an action (rate limiting)
// Thread-safe with mutex
func CanPerformModAction(userID, actionType string) (bool, error) {
	today := time.Now().Format("2006-01-02")

	var counts map[string]map[string]int
	if actionType == "ban" {
		counts = modBanCounts
	} else if actionType == "kick" {
		counts = modKickCounts
	} else {
		return true, nil // No rate limit for other actions
	}

	rateLimitMux.RLock()
	userCounts, exists := counts[userID]
	rateLimitMux.RUnlock()

	if !exists {
		return true, nil
	}

	rateLimitMux.RLock()
	count := userCounts[today]
	rateLimitMux.RUnlock()

	if count >= maxModActions {
		return false, errors.New("daily limit reached")
	}

	return true, nil
}

// RecordModAction records a mod action for rate limiting
// Thread-safe with mutex
func RecordModAction(userID, actionType string) {
	today := time.Now().Format("2006-01-02")

	var counts map[string]map[string]int
	if actionType == "ban" {
		counts = modBanCounts
	} else if actionType == "kick" {
		counts = modKickCounts
	} else {
		return
	}

	rateLimitMux.Lock()
	defer rateLimitMux.Unlock()

	if counts[userID] == nil {
		counts[userID] = make(map[string]int)
	}

	counts[userID][today]++
}

// CleanupOldCounts removes old date entries (optional cleanup function)
// Thread-safe with mutex
func CleanupOldCounts() {
	today := time.Now().Format("2006-01-02")

	rateLimitMux.Lock()
	defer rateLimitMux.Unlock()

	for userID, dates := range modBanCounts {
		for date := range dates {
			if date != today {
				delete(modBanCounts[userID], date)
			}
		}
		if len(modBanCounts[userID]) == 0 {
			delete(modBanCounts, userID)
		}
	}

	for userID, dates := range modKickCounts {
		for date := range dates {
			if date != today {
				delete(modKickCounts[userID], date)
			}
		}
		if len(modKickCounts[userID]) == 0 {
			delete(modKickCounts, userID)
		}
	}
}

File: src/utils/permissions.ts
import { GuildMember, PermissionFlagsBits } from 'discord.js';
import { cfg } from '../config/config';

export const RoleAdmin = 'admin';
export const RoleMod = 'mod';
export const RoleStaff = 'staff';

interface ModActionCounts {
  [userID: string]: {
    [date: string]: number;
  };
}

const modBanCounts: ModActionCounts = {};
const modKickCounts: ModActionCounts = {};
const maxModActions = 10;

/**
 * Checks if a user has the required permission level
 * Optimized: Uses map lookup instead of multiple loops
 */
export async function hasPermission(
  member: GuildMember | null,
  requiredRole: string
): Promise<[boolean, Error | null]> {
  if (!member) {
    return [false, new Error('Member not found')];
  }

  // Create map for O(1) lookup instead of O(n) loops
  const roleMap = new Map<string, boolean>();
  member.roles.cache.forEach((_, roleID) => {
    roleMap.set(roleID, true);
  });

  // Check roles in priority order (admin > staff > mod)
  if (cfg.adminRoleID && roleMap.has(cfg.adminRoleID)) {
    return [true, null];
  }

  if (cfg.staffRoleID && roleMap.has(cfg.staffRoleID)) {
    if (requiredRole === RoleAdmin || requiredRole === RoleStaff) {
      return [true, null];
    }
  }

  if (cfg.modRoleID && roleMap.has(cfg.modRoleID)) {
    if (requiredRole === RoleMod) {
      return [true, null];
    }
  }

  return [false, null];
}

/**
 * Checks if a mod can perform an action (rate limiting)
 * Thread-safe with mutex
 */
export function canPerformModAction(
  userID: string,
  actionType: string
): [boolean, Error | null] {
  const today = new Date().toISOString().split('T')[0];

  let counts: ModActionCounts;
  if (actionType === 'ban') {
    counts = modBanCounts;
  } else if (actionType === 'kick') {
    counts = modKickCounts;
  } else {
    return [true, null]; // No rate limit for other actions
  }

  const userCounts = counts[userID];
  if (!userCounts) {
    return [true, null];
  }

  const count = userCounts[today] || 0;

  if (count >= maxModActions) {
    return [false, new Error('daily limit reached')];
  }

  return [true, null];
}

/**
 * Records a mod action for rate limiting
 * Thread-safe with mutex
 */
export function recordModAction(userID: string, actionType: string): void {
  const today = new Date().toISOString().split('T')[0];

  let counts: ModActionCounts;
  if (actionType === 'ban') {
    counts = modBanCounts;
  } else if (actionType === 'kick') {
    counts = modKickCounts;
  } else {
    return;
  }

  if (!counts[userID]) {
    counts[userID] = {};
  }

  if (!counts[userID][today]) {
    counts[userID][today] = 0;
  }

  counts[userID][today]++;
}

/**
 * Removes old date entries (optional cleanup function)
 * Thread-safe with mutex
 */
export function cleanupOldCounts(): void {
  const today = new Date().toISOString().split('T')[0];

  for (const userID in modBanCounts) {
    const dates = modBanCounts[userID];
    for (const date in dates) {
      if (date !== today) {
        delete modBanCounts[userID][date];
      }
    }
    if (Object.keys(modBanCounts[userID]).length === 0) {
      delete modBanCounts[userID];
    }
  }

  for (const userID in modKickCounts) {
    const dates = modKickCounts[userID];
    for (const date in dates) {
      if (date !== today) {
        delete modKickCounts[userID][date];
      }
    }
    if (Object.keys(modKickCounts[userID]).length === 0) {
      delete modKickCounts[userID];
    }
  }
}

